From 70ab6ca18864d0511e2f6cea1d12be58ee926c81 Mon Sep 17 00:00:00 2001
From: Aditya Kumar Singh <quic_adisi@quicinc.com>
Date: Wed, 12 Apr 2023 21:25:43 +0530
Subject: [PATCH] ath12k: enable sending TPC power command for 320 MHz

TPE spec is not yet defined for 320 MHz. However, for proper
Tx power related calculations, it is mandatory to send SET_TPC_POWER
command to firmware in 320 MHz as well.

Add support to form and send the TPC power command for 320 MHz bandwidth
as well.

Signed-off-by: Aditya Kumar Singh <quic_adisi@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 9 +++++++++++++++
 include/linux/ieee80211.h | 2 +-
 2 files changed, 10 insertions(+), 1 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -7433,6 +7433,9 @@ static u8 ath12k_mac_get_num_pwr_levels(
                 case NL80211_CHAN_WIDTH_160:
                         num_pwr_levels = 8;
                         break;
+                case NL80211_CHAN_WIDTH_320:
+                	num_pwr_levels = 16;
+                	break;
                 default:
                         return 1;
                 }
@@ -7451,6 +7454,9 @@ static u8 ath12k_mac_get_num_pwr_levels(
                 case NL80211_CHAN_WIDTH_160:
                         num_pwr_levels = 4;
                         break;
+                case NL80211_CHAN_WIDTH_320:
+                	num_pwr_levels = 5;
+                	break;
                 default:
                         return 1;
                 }
@@ -7473,6 +7479,9 @@ static u16 ath12k_mac_get_6g_start_frequ
          * with center frequency 5955, its diff is 6025 - 5955 = 70.
          */
         switch (chan_def->width) {
+        case NL80211_CHAN_WIDTH_320:
+        	diff_seq = 150;
+        	break;
         case NL80211_CHAN_WIDTH_160:
                 diff_seq = 70;
                 break;
